### ISSUE TITLE
Refactor report JSON exports

### ISSUE THREAD

**Issue by max-ostapenko** [HTTPArchive/httparchive.org#938]
For the script located in https://github.com/HTTPArchive/bigquery/tree/master/sql:

- [x] - rewrite queries to a new stable schema
- [x] - move the script to another currently used repo where it fits the most
- [x] - trigger this step automatically upon successful BQ pipeline completion
- [x] - Handle CrUX reports in addition to crawl

I can help with SQL rewrite, but not familiar with nuances of these bash scripts

Related issues:
- documentation https://github.com/HTTPArchive/data-pipeline/issues/21
- alerts for failed reports (already implemented for failed workflows on Dataform) https://github.com/HTTPArchive/data-pipeline/issues/24

**Comment by tunetheweb** [comment]
They've been hacked and hacked and hacked and are now a mess. Mostly my hacking. Open to better ideas here (dataform?).

Basically they need to be able to do this:
- Run monthly histograms SQLs when crawl is finished
- Run longer term time series SQLs when crawl is finished
  - Be able to run the time series in an incremental fashion (download existing JSON, run for any months since then, and merge them together—this lead to HUGE timing savings) which currently requires injecting additional SQL clauses (using different data formats if it's HTTP Archive legacy, `all`, or CrUX.
- Handle different lenses (Top X, WordPress, Drupal, Magento) which currently requires injecting additional SQL clauses into the current queries. We should make this natively supported in the current SQL with an optional param.
- Handle CrUX reports (monthly histograms and time series) having to run later.
- Be able to upload to cloud storage in GCP to allow it to be hosted on our CDN ([example json](https://cdn.httparchive.org/reports/numUrls.json)) and be served to our site ([example report](https://httparchive.org/reports/state-of-the-web)) or others.
- Be able to run and only run reports missing (histograms) or missing dates (time series)
- Be able to force rerun (to override any existing reports). Note that for time series this means running from beginning of time so can take a long time (especially for some of the expensive ones!), but we don't do that often.
- Be able to run a subset of reports (e.g. all `*crux*` ones, or just the `numURLs` one casue it was wrong and needs force rerun).

Thoughts on how best to tackle this WITHOUT hacky bash scripts?

**Comment by max-ostapenko** [comment]
Yeah, seems Dataform matches requirements well (simple and native)

So I suppose it's a manual step today?
Deployed somewhere or running locally?
How much time does it take?

**Comment by tunetheweb** [comment]
Deployed on a VM in the cloud. We do a "git pull" to release a new report and it's kicked off from cron every morning and bails early if tables not ready, and then runs each report when data is available, and then checks (but does not run) each report every other day of that month. A bit wasteful to be honest!

It takes a full day to run all reports for all lenses.

**Comment by tunetheweb** [comment]
> Missing why?

Ideally no missing data. But sometimes it fails and I don't get run to fixing it that month. The way it works currently is it checks the last date and then runs the report with `AND date > LAST_GOOD_DATE` which means it automatically fills in the missing data for the time series without me having to run it twice for each month. Not essential but kinda nice feature of the current script.

**Comment by max-ostapenko** [comment]
Oh, so it's the queries that I see running today (with VM service account)
Based on [logs](https://cloudlogging.app.goo.gl/EiQ9Vc5p3jjpC4rp7) it takes more than 2 days ([PWA scores](https://cdn.httparchive.org/reports/top1k/pwaScores.json) not available while running) and 17% of jobs fail.

And looking at [first histogram report](https://github.com/HTTPArchive/bigquery/blob/master/sql/histograms/bootupJs.sql) bytes estimate we are doing 35X more processing that would be required in case of incremental table.

I would focus on rewriting queries to new `all` schema so that we can deprecate legacy.
And if you think it will break the current process and can help rewriting bash into js templates - we can go full migration.

**Comment by max-ostapenko** [comment]
BTW if I understood correctly the current solution is constantly checking and running missing reports without much manual interference today - it's not the case with Dataform.

When jobs fail in Dataform we'll need to fix and re-run them manually (hoping next time the DAG will work better). Queuing reports and retrying failed later doesn't work with this tool.

**Comment by tunetheweb** [comment]
> Oh, so it's the queries that I see running today (with VM service account)

Yup.

> Based on [logs](https://cloudlogging.app.goo.gl/EiQ9Vc5p3jjpC4rp7) it takes more than 2 days ([PWA scores](https://cdn.httparchive.org/reports/top1k/pwaScores.json) not available while running)

Oh that's a good point! That no longer exists so will never complete. But it's running everyday trying. Have deleted that file so it won't be run anymore. Will delete from GitHub too.

> and 17% of jobs fail.

As well as odd examples like above, CrUX ones will fail until that data is ready. And some don't work with lenses (but we don't have a good way of excluding them). As those will run everyday (without success) that probably inflates the failure rates. But yeah would be nice to be able to schedule these better so we only have real failures.

> And looking at [first histogram report](https://github.com/HTTPArchive/bigquery/blob/master/sql/histograms/bootupJs.sql) bytes estimate we are doing 35X more processing that would be required in case of incremental table.

This is because the Lighthouse report will now be JSON and so don't need to process the whole report to get a very small subset that data? That will indeed be a very nice improvement!

> 
> I would focus on rewriting queries to new `all` schema so that we can deprecate legacy. And if you think it will break the current process and can help rewriting bash into js templates - we can go full migration.

Yeah I started on that. Will try to finish it out.

> BTW if I understood correctly the current solution is constantly checking and trying to process missing reports without much manual interference today - it's not the case with Dataform.

Yes. But that also has downsides as per above! It was very low tech. We can do better here so we don't need to do that.

> When jobs fail in Dataform we'll need to fix and re-run them manually (the correct DAG being the solution). Queuing reports and retrying failed later doesn't work with this tool.

That shouldn't be an issue if we make this more robust.

**Comment by tunetheweb** [comment]
The reports in the `/histograms/` folder only run for the month needed: `${YYYY_MM_DD}` is replaced by the actual date in the bash script.

**Comment by max-ostapenko** [comment]
Oh, true. I was talking about time series: https://github.com/HTTPArchive/bigquery/blob/master/sql/timeseries/asyncClipboardRead.sql

**Comment by max-ostapenko** [comment]
Oh, I expected something like this.
Looks like magic unless you understand the script...

Ok, let me create a branch with some drafts, so that you can get a feeling about the workflow and templating.

**Comment by tunetheweb** [comment]
Yeah the script started simple, and then grew and grew and grew - well beyond what we should use a bash script for!

But hey, it works!

**Comment by max-ostapenko** [comment]
@tunetheweb I expect to merge https://github.com/HTTPArchive/dataform/pull/30 so that we can deliver migration and improvements to the tomorrow's CWV Tech prod pipeline run.

GCS JSON reports are currently configured with demo 1 timeseries and 1 histogram published to `httparchive/reports/dev`.

JSON reports will be ready for prod once:
- the rest of metrics SQLs added to [reports_config.js](https://github.com/HTTPArchive/dataform/blob/reports/includes/reports_config.js)
- lenses logic is added

**Comment by tunetheweb** [comment]
@max-ostapenko I think we can close this yes?

**Comment by max-ostapenko** [comment]
We tackled only the migration to the `crawl` dataset.
There is a big topic of reports processing orchestration that is still open, and sits in a half-baked branch.

I'd like to have stricter control over what is run, when and the export options.
Eventually I expect we orchestrate these similarly to tech reports.

Should I rename it?

**Comment by max-ostapenko** [comment]
@tunetheweb I hope we've pulled it off.
You removed the cron so in 3 weeks we'll see how successful it is.
